### PR TITLE
WW-4930 Add constant to enable smi inheritance

### PR DESCRIPTION
**Pull request by pHeidelbach**
Adds the `struts.convention.enable.smi.inheritance` constant which determines if a newly created package config inherits the strict-method-invocation value of its parent package config.

### CODE REVIEW

**Comment by pHeidelbach** [comment]
You're right. I will create one.

**Comment by pHeidelbach** [comment]
I added the corresponding test.

File: plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java [hunk 1]
```diff
@@ -121,6 +134,19 @@ private void run(String actionPackages, String packageLocators, String excludePa
         InterceptorMapping interceptor2 = new InterceptorMapping("interceptor-2", new TestInterceptor());
         defaultInterceptorStacks.add(makeInterceptorStackConfig("stack-1", interceptor1, interceptor2));
 
+      //setup strict MethodInvocation
```
Comment (sepe81) [inline plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java]: fix indentation

File: plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java [hunk 2]
```diff
@@ -121,6 +134,19 @@ private void run(String actionPackages, String packageLocators, String excludePa
...
+        boolean strictMethodInvocation = true;
+        boolean isSmiInheritanceEnabled = false;
+        // Sets the SMI value to false so we can test that the created package configs adopt the SMI value of their parent.
+        // If enableSmiInheritance is set to false it is expected that the generated package configs have their SMI value set to true (the default)
+        // even when the SMI of their parent is set to false.
+        if(StringUtils.equals(enableSmiInheritance, "true")){
+            strictMethodInvocation = false;
+            isSmiInheritanceEnabled = true;
+        }else if (StringUtils.equals(enableSmiInheritance, "false")) {
```
Comment (sepe81) [inline plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java]: fix formatting at `}else if`

File: plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java [hunk 3]
```diff
@@ -539,37 +578,50 @@ public Container getContainer() {
...
         assertTrue(actionConfig.getAllowedMethods().contains("execute"));
-        assertTrue(actionConfig.getAllowedMethods().contains("end"));
-        assertTrue(actionConfig.getAllowedMethods().contains("input"));
-        assertTrue(actionConfig.getAllowedMethods().contains("cancel"));
-        assertTrue(actionConfig.getAllowedMethods().contains("start"));
-        assertTrue(actionConfig.getAllowedMethods().contains("home"));
-        assertTrue(actionConfig.getAllowedMethods().contains("browse"));
+        if(isSmiInheritanceEnabled) {
+            // With strict method invocation disabled the allowed methods are "execute" and the wildcard "*"
+            assertEquals(2, actionConfig.getAllowedMethods().size());
```
Comment (sepe81) [inline plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java]: Maybe you could extract `actionConfig.getAllowedMethods().size()` and put it before the if stmt?

File: plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java [hunk 4]
```diff
@@ -539,37 +578,50 @@ public Container getContainer() {
...
         assertEquals("struts-default", pkgConfig.getParents().get(0).getName());
 
         actionConfig = pkgConfig.getActionConfigs().get("package-level-allowed-methods-child");
-        assertEquals(actionConfig.getAllowedMethods().size(), 6);
         assertTrue(actionConfig.getAllowedMethods().contains("execute"));
+        if(isSmiInheritanceEnabled) {
+            // With strict method invocation disabled the allowed methods are execute and the wildcard *
+            assertEquals(2, actionConfig.getAllowedMethods().size());
+        } else {
+        assertEquals(6, actionConfig.getAllowedMethods().size());
```
Comment (sepe81) [inline plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java]: fix indentation

Comment (sepe81) [reply plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java]: ...maybe even for the existing 3 lines

**Comment by sepe81** [comment]
LGTM, just some minor tasks for TODO and hopefully the build stays green 👍

**Comment by pHeidelbach** [comment]
Added a fixup to resolve the issues.

File: plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java [hunk 5]
```diff
@@ -121,6 +134,19 @@ private void run(String actionPackages, String packageLocators, String excludePa
         InterceptorMapping interceptor2 = new InterceptorMapping("interceptor-2", new TestInterceptor());
         defaultInterceptorStacks.add(makeInterceptorStackConfig("stack-1", interceptor1, interceptor2));
 
+        //setup strict MethodInvocation
+        boolean strictMethodInvocation = true;
+        boolean isSmiInheritanceEnabled = false;
+        // Sets the SMI value to false so we can test that the created package configs adopt the SMI value of their parent.
+        // If enableSmiInheritance is set to false it is expected that the generated package configs have their SMI value set to true (the default)
+        // even when the SMI of their parent is set to false.
+        if(StringUtils.equals(enableSmiInheritance, "true")){
```
Comment (lukaszlenart) [inline plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java]:
```suggestion
        if (StringUtils.equals(enableSmiInheritance, "true")) {
```

File: plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java [hunk 6]
```diff
@@ -347,7 +372,7 @@ public Container getContainer() {
         mockContainer.setResultMapBuilder(resultMapBuilder);
         mockContainer.setConventionsService(new ConventionsServiceImpl(""));
 
-        PackageBasedActionConfigBuilder builder = new PackageBasedActionConfigBuilder(configuration, mockContainer ,of, "false", "struts-default");
+        PackageBasedActionConfigBuilder builder = new PackageBasedActionConfigBuilder(configuration, mockContainer ,of, "false", "struts-default", enableSmiInheritance);
```
Comment (lukaszlenart) [inline plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java]:
```suggestion
        PackageBasedActionConfigBuilder builder = new PackageBasedActionConfigBuilder(configuration, mockContainer, of, "false", "struts-default", enableSmiInheritance);
```

File: plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java [hunk 7]
```diff
@@ -539,37 +578,52 @@ public Container getContainer() {
...
-        assertEquals(7, actionConfig.getAllowedMethods().size());
         assertTrue(actionConfig.getAllowedMethods().contains("execute"));
-        assertTrue(actionConfig.getAllowedMethods().contains("end"));
-        assertTrue(actionConfig.getAllowedMethods().contains("input"));
-        assertTrue(actionConfig.getAllowedMethods().contains("cancel"));
-        assertTrue(actionConfig.getAllowedMethods().contains("start"));
-        assertTrue(actionConfig.getAllowedMethods().contains("home"));
-        assertTrue(actionConfig.getAllowedMethods().contains("browse"));
+        int allowedMethodsSize = actionConfig.getAllowedMethods().size();
+        if(!pkgConfig.isStrictMethodInvocation()) {
```
Comment (lukaszlenart) [inline plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java]:
```suggestion
        if (!pkgConfig.isStrictMethodInvocation()) {
```

File: plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java [hunk 8]
```diff
@@ -539,37 +578,52 @@ public Container getContainer() {
...
         assertEquals("struts-default", pkgConfig.getParents().get(0).getName());
 
         actionConfig = pkgConfig.getActionConfigs().get("package-level-allowed-methods-child");
-        assertEquals(actionConfig.getAllowedMethods().size(), 6);
         assertTrue(actionConfig.getAllowedMethods().contains("execute"));
-        assertTrue(actionConfig.getAllowedMethods().contains("home"));
-        assertTrue(actionConfig.getAllowedMethods().contains("start"));
-        assertTrue(actionConfig.getAllowedMethods().contains("input"));
+        allowedMethodsSize = actionConfig.getAllowedMethods().size();
+        if(!pkgConfig.isStrictMethodInvocation()) {
```
Comment (lukaszlenart) [inline plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java]:
```suggestion
        if (!pkgConfig.isStrictMethodInvocation()) {
```

File: plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java [hunk 9]
```diff
@@ -672,6 +731,14 @@ private void verifyActionConfig(PackageConfig pkgConfig, String actionName, Stri
...
     }
 
+    private void checkSmiValue(PackageConfig pkgConfig, PackageConfig parentConfig,  boolean isSmiInheritanceEnabled) {
+        if (isSmiInheritanceEnabled) {
+            assertEquals(parentConfig.isStrictMethodInvocation(), pkgConfig.isStrictMethodInvocation());
+        } else if (!isSmiInheritanceEnabled && !parentConfig.isStrictMethodInvocation()){
+            assertTrue(pkgConfig.isStrictMethodInvocation());
+        }
+    }
+
```
Comment (lukaszlenart) [inline plugins/convention/src/test/java/org/apache/struts2/convention/PackageBasedActionConfigBuilderTest.java]: 👏

**Comment by pHeidelbach** [comment]
Fixed the style issues and squashed the fixup.

**Comment by lukaszlenart** [comment]
Let's rock 👍

**Pull request merged**